### PR TITLE
Shipa 2720

### PR DESCRIPTION
**Pull request by kavinaravind**
# Description

This PR adds Ketch CLI support for applications to bind to a volume as well as configuring a pod level security context in regards to `fsGroup` and `runAsUser`

for the following PV and PVC:

```
apiVersion: v1
kind: PersistentVolume
metadata:
  name: pv-volume-1
  labels:
    type: local
spec:
  storageClassName: manual
  capacity:
    storage: 10Gi
  accessModes:
    - ReadWriteOnce
  hostPath:
    path: "/mnt/data"
---
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: pvc-1
spec:
  storageClassName: manual
  accessModes:
    - ReadWriteOnce
  resources:
    requests:
      storage: 3Gi
```

The following ketch command will bind to this pvc as well as set the security context accordingly

```
ketch/. app deploy bulletinboard -i docker.io/shipasoftware/bulletinboard:1.0 -k dev --volume pvc-1 --volume-mount-path /var/lib/sql --fs-group 1 --volume-mount-options readOnly=true --run-as-user 0
```

Another addition is that during the first deployment and if a volume is specified, a StatefulSet will be created rather than a Deployment.

Fixes # (issue)

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Chore (documentation addition or typo, file relocation)

## Testing

- [x] New tests were added with this PR that prove my fix is effective or that my feature works (describe below this bullet)
- [ ] This change requires no testing (i.e. documentation update)

## Documentation

- [ ] All added public packages, funcs, and types have been documented with doc comments
- [ ] I have commented my code, particularly in hard-to-understand areas

## Final Checklist:

- [ ] I followed standard [GitHub flow](https://guides.github.com/introduction/flow/) guidelines
- [ ] I have performed a self-review of my own code
- [ ] My changes generate no new warnings

## Additional Information (omit if empty)

Please include anything else relevant to this PR that may be useful to know.

### CODE REVIEW

File: internal/deploy/parameters.go [hunk 1]
```diff
@@ -363,6 +396,89 @@ func (c *ChangeSet) getProcess() (string, error) {
...
+			if err != nil {
+				return nil, fmt.Errorf("%w readOnly must be either true or false",
+					newInvalidUsageError(FlagVolumeMountOptions))
+			}
+			volumeMount.ReadOnly = parsed
+		}
+	}
+
+	// if empty, no mount paths or options were set
+	if volumeMount == (v1.VolumeMount{}) {
```
Comment (stinkyfingers) [inline internal/deploy/parameters.go:452]: Ha, makes sense, but a little unorthodox.

File: internal/deploy/deploy.go [hunk 2]
```diff
@@ -348,6 +387,14 @@ func updateAppCRD(ctx context.Context, svc *Services, appName string, args updat
 				Cmd:  cmd,
 			}
 
+			if args.process == "" {
```
Comment (stinkyfingers) [inline internal/deploy/deploy.go:390]: Condense into `if args.process == "" || args.process == processName`?

File: internal/deploy/parameters_test.go [hunk 3]
```diff
@@ -48,3 +55,258 @@ func TestChangeSet_getStepWeight(t *testing.T) {
...
+		want    []v1.Volume
+		wantErr string
+	}{
+		{
+			name:    "volume name not included",
+			set:     ChangeSet{},
+			wantErr: `"volume" missing`,
+		},
+		{
+			name: "no volume source",
```
Comment (stinkyfingers) [inline internal/deploy/parameters_test.go:109]: Do tests `no volume source`and `incorrect volume source` test the same thing as `valid volume`?

File: internal/deploy/parameters_test.go [hunk 4]
```diff
@@ -48,3 +55,258 @@ func TestChangeSet_getStepWeight(t *testing.T) {
...
+			set:     ChangeSet{volumeMountOptions: &map[string]string{"readOnly": "true"}},
+			wantErr: `"volume-mount-options" used improperly volume-mount-options must be used with volume flag`,
+		},
+		{
+			name:    "volume mount path set with invalid readOnly value",
+			set:     ChangeSet{volume: stringRef("pvc-1"), volumeMountOptions: &map[string]string{"readOnly": "asdf"}},
+			wantErr: `"volume-mount-options" used improperly readOnly must be either true or false`,
+		},
+		{
+			name:    "volume mount path set with invalid readOnly value",
```
Comment (stinkyfingers) [inline internal/deploy/parameters_test.go:197]: Duplicate of above test?

**Review by stinkyfingers — COMMENTED** [review]
Looks good. Just a couple small questions/comments.

File: internal/deploy/parameters.go (hunk 1 above)
Comment (kavinaravind) [reply internal/deploy/parameters.go:452]: very true, think its okay for this case? I didnt feel the need to check every field within this struct and rather used this short hand (albeit kinda unconventional)

File: internal/deploy/parameters_test.go (hunk 3 above)
Comment (kavinaravind) [reply internal/deploy/parameters_test.go:109]: `valid volume` will check specifically for the existence of VolumeSource of type PersistentVolumeClaim while the other two error tests either lack PersistentVolumeClaim (its a ConfigMap instead) or where PersistentVolumeClaim does not exist

**Review by stinkyfingers — APPROVED** [review]
Nice work! I created an app with volumes, as-user , etc. successfully.

**Pull request merged**